### PR TITLE
Don't break mypy cache/incremental mode

### LINKED ISSUE #4: Cannot find component `ProtocolIntersection\u200b`
If one module imports `ProtocolIntersection`, and then another module attempts to import it from that module, `mypy` seems unable to lookup `ProtocolIntersection` in the imported module's namespace (maybe due to the presence of a zero-width space at the end?):
```python
Traceback (most recent call last):
  File "C:\Users\evanj\GitHub\coconut\coconut\command\mypy.py", line 69, in mypy_run
    stdout, stderr, exit_code = run(args)
  File "mypy\api.py", line 72, in run
  File "mypy\api.py", line 59, in _run
  File "mypy\api.py", line 73, in <lambda>
  File "mypy\main.py", line 95, in main
  File "mypy\main.py", line 174, in run_build
  File "mypy\build.py", line 194, in build
  File "mypy\build.py", line 277, in _build
  File "mypy\build.py", line 2930, in dispatch
  File "mypy\build.py", line 3321, in process_graph
  File "mypy\build.py", line 3402, in process_fresh_modules
  File "mypy\build.py", line 2121, in fix_cross_refs
  File "mypy\fixup.py", line 53, in fixup_module
  File "mypy\fixup.py", line 136, in visit_symbol_table
  File "mypy\fixup.py", line 73, in visit_type_info
  File "mypy\fixup.py", line 138, in visit_symbol_table
  File "mypy\nodes.py", line 774, in accept
  File "mypy\fixup.py", line 146, in visit_func_def
  File "mypy\types.py", line 1869, in accept
  File "mypy\fixup.py", line 244, in visit_callable_type
  File "mypy\types.py", line 2803, in accept
  File "mypy\fixup.py", line 338, in visit_type_type
  File "mypy\types.py", line 1357, in accept
  File "mypy\fixup.py", line 209, in visit_instance
  File "mypy\fixup.py", line 344, in lookup_fully_qualified_typeinfo
  File "mypy\lookup.py", line 49, in lookup_fully_qualified
AssertionError: Cannot find component 'ProtocolIntersection\u200b' for 'typing_protocol_intersection.types.ProtocolIntersection\u200b'
```

Also note that reproduction here is a bit tricky, since it seems to depend on the current status of `.mypy_cache`, though it always shows up for me after running `mypy` enough times. I'm on Python 3.9 and `mypy` 1.2.0. `mypy.ini` is just:
```ini
[mypy]
plugins = typing_protocol_intersection.mypy_plugin
```

### PR DESCRIPTION
**Pull request by drvink**
Fixes #4.

### CODE REVIEW

**Comment by klausweiss** [comment]
Hey. Thanks a lot for this! I'll try to dive into this in the next few days.

File: typing_protocol_intersection/mypy_plugin.py [hunk 1]
```diff
@@ -23,6 +23,19 @@
 class ProtocolIntersectionPlugin(mypy.plugin.Plugin):
     # pylint: disable=unused-argument
 
+    def report_config_data(
```
Comment (klausweiss) [inline typing_protocol_intersection/mypy_plugin.py:26]: Note to self: see https://mypy.readthedocs.io/en/stable/extending_mypy.html?highlight=report_config_data#current-list-of-plugin-hooks

**Review by klausweiss — COMMENTED** [review]
Super cool you figured this out! Thanks for your contribution. Any chance for an automated test that would break prior to this change, but doesn't now? Something along the lines of [this gist](https://github.com/klausweiss/typing-protocol-intersection/issues/4#issuecomment-1518905366) from #4 would do.

**Comment by drvink** [comment]
If that gist used to reproduce the issue and no longer does, isn't that sufficient for the test? (You can push changes to this PR yourself to add a test before you merge it; probably easier for you to do it than me.)

File: typing_protocol_intersection/mypy_plugin.py [hunk 2]
```diff
@@ -23,6 +23,19 @@
...
+        self, ctx: mypy.plugin.ReportConfigContext
+    ) -> bool:
+        # When is_check is False, the value returned by this function will be
+        # serialized as json and saved to the mypy cache.  When is_check is
+        # True, the value returned will be compared against whatever was loaded
+        # from the cache.  This is so that a plugin can potentially invalidate
```
[the remark this answers is not in the document]
Comment (drvink) [reply typing_protocol_intersection/mypy_plugin.py:32]: The callback is first invoked pre-run (`ctx.is_check` == True) in order to check that the loaded cache contents match the plugin's desired state, and later invoked post-run (`ctx.is_check` == False) when mypy asks the plugin what it wants to store to the cache, if anything.

File: typing_protocol_intersection/mypy_plugin.py [hunk 3]
```diff
@@ -23,6 +23,19 @@
...
+        # the cache for a module, which is exactly what we want to do if we
+        # generated any types on a previous run; otherwise mypy will try to look
+        # them up in the cache (and subsequently crash when it fails to find
+        # them).
+        return UniqueFullname.instance_counter == 0
```
[the remark this answers is not in the document]
Comment (drvink) [reply typing_protocol_intersection/mypy_plugin.py:37]: As far as I can tell, the situation is simply that if the funny unique names end up in the cache, mypy breaks. So it doesn't matter exactly how many were generated; we simply want to ensure that the cache for a given file is invalidated if the plugin generated any unique names during a previous typechecking run for that file. We don't care about the case where there's no cached plugin state, and we also don't need to worry about files that didn't get funny names written into their cache, so as long as we get the caches invalidated for ones that did, everything is fine.

You can see how this dumb trick works for yourself by using this:

```python
        ret = UniqueFullname.instance_counter == 0
        print(f'[{ctx.is_check}] {ctx.path}: {ret}', file=sys.stderr)
        return ret 
```

Or, if you're feeling formal, I think this translation is correct:

```
∀ isCheck cachedVal didNoWork,
  if isCheck ∧ cachedVal then didNoWork = cachedVal ⇒ didNoWork else
  if isCheck ∧ ¬cachedVal then didNoWork ≠ cachedVal ⇒ didNoWork else
  if ¬isCheck then didNoWork ∨ ¬didNoWork else
  ⊥
```

which works out to be true.

**Comment by klausweiss** [comment]
> You can push changes to this PR yourself to add a test before you merge it; probably easier for you to do it than me.

It probably is, sure. It'll probably take a few days for me to get round to it.

> [edit] I just tried the gist and it makes mypy error, because the plugin doesn't have the right stuff to support using ProtocolIntersection[...] as a type alias, only in function signatures and class inheritance lists. I do have a file that triggers this bug pre-PR if you want it, but I don't have time to add a test for it.

If you have one which makes mypy pass and used to crash before your change, that would be great, thanks! I'll try to find the time to wrap it in an integration test. Good call with the type alias support BTW. I'll open an issue when I investigate a bit more. Anyone is free to open if you beat me to it ;)

**Comment by drvink** [comment]
> If you have one which makes mypy pass and used to crash before your change, that would be great, thanks! I'll try to find the time to wrap it in an integration test. Good call with the type alias support BTW. I'll open an issue when I investigate a bit more. Anyone is free to open if you beat me to it ;)

[This](https://github.com/klausweiss/typing-protocol-intersection/files/12443304/descriptor.txt) should do it.

**Comment by klausweiss** [comment]
I couldn't reproduce the original error with this file under python 3.11 and mypy 1.15. Are you sure this is the right file? It doesn't import `typing_protocol_intersection`.

Anyway, I have finally found some time and utilized the script from the original gist for an automated test case (FYI @evhub, and many thanks!). I have modified `report_config_data` to hopefully return a value more in the spirit of what it's supposed to do. Some black scramble to resolve and I'll merge.

**Pull request merged**